### PR TITLE
Add Tournament Result Logging

### PR DESCRIPTION
**Pull request by r3dDoX**
Merge branch 'develop' into feature/tournament-logging

### CODE REVIEW

File: server/communication/jsonResultProxy.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    if (serverMessageType === MessageType.REQUEST_TRUMPF.name) {
+        return MessageType.CHOOSE_TRUMPF.name;
+    }
+    return MessageType.CHOOSE_CARD.name;
+};
+
+const JsonResultProxy = {
+
+    destroy() {
+        this.fileWriter.write(']', (err) => this.fileWriter.close());
```
Comment (danielsuter) [inline server/communication/jsonResultProxy.js]: err is not used

File: server/communication/jsonResultProxy.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    get(target, propKey) {
+        return (...args) => {
+            const result = target[propKey].apply(target, args);
+
+            if (methodsToIntercept.includes(propKey)) {
+                const clientId = args.shift().jassChallengeId;
+
+                const serverSendResult = {
+                    broadcast: !clientId,
```
Comment (danielsuter) [inline server/communication/jsonResultProxy.js]: Add comment why this works

File: server/communication/jsonResultProxy.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                    sentTo: clientId,
+                    messageType: args.shift(),
+                    data: args.pop()
+                };
+
+                this.writeResult(serverSendResult);
+
+                if (result && typeof result.then === 'function') {
```
Comment (danielsuter) [inline server/communication/jsonResultProxy.js]: extract method isPromise(...)

File: server/communication/clientApi.js [hunk 4]
```diff
@@ -147,12 +147,17 @@ const ClientApi = {
     },
 
     subscribeMessage(client, messageType, messageHandler) {
-        return ClientCommunication.on(client, messageType, messageHandler);
+        return this.clientCommunication.on(client, messageType, messageHandler);
+    },
+
+    registerCommunicationProxy(resultWriter) {
+        this.clientCommunication = new Proxy(ClientCommunication, resultWriter);
     }
```
Comment (r3dDoX) [inline server/communication/clientApi.js]: done 👍

File: server/communication/jsonResultProxy.js (hunk 1 above)
Comment (r3dDoX) [inline server/communication/jsonResultProxy.js]: done 👍

File: server/communication/jsonResultProxy.js (hunk 2 above)
Comment (r3dDoX) [inline server/communication/jsonResultProxy.js]: done 👍

File: server/communication/jsonResultProxy.js (hunk 3 above)
Comment (r3dDoX) [inline server/communication/jsonResultProxy.js]: done :+1:

**Comment by r3dDoX** [comment]
All review comments implemented. Please review again

**Pull request merged**